### ISSUE TITLE
Vertical scroll should keep constant visible range in log-Hz views, instead of preserving range in Hz

### ISSUE THREAD

**Issue by St3v3C** [sonic-visualiser/sonic-visualiser#14]
Key and Mouse Ref indicates ALT + Wheel should scroll vertically instead it scrolls horizontally the same as CTRL + Wheel. The only way to move vertically is to Shift + Left Drag to Zoom to an area. Sonic Visualizer Version 4.3 on Windows 10. This isn't a perfect work around as you also get unintended zoom when all you want is vertical scroll.

**Comment by cannam** [comment]
I can confirm this, thank you for the report.

However, it might not be all that bad - I've checked the code, and it looks as if the docs (i.e. the key and mouse reference) simply list Shift+Wheel and Alt+Wheel the wrong way around. Try Shift+Wheel to scroll vertically and Alt+Wheel to zoom.

I'm not sure whether it was always this way, or whether at some point the code has changed (but why?) to invalidate the docs. I suppose I could dig in and find out.

If you can confirm this works for you, then I'll go ahead and simply correct the docs for the next release.

(Most of the key combinations listed in Key and Mouse Reference are taken automatically from the actual menu shortcuts, and so are "guaranteed" to be current and correct - but the general pane navigation ones sadly are not)

**Comment by St3v3C** [comment]
Hmm, something seems not right. For me...Shift + Wheel     Seems to do vertical Zoom and Scroll at the same timeAlt + Wheel        Does horizontal scroll (duplicates Ctrl + Wheel)
So I can't seem to do vertical scroll at a constant zoom level.
The same if I do Left Drag in the vertical direction. it zooms and scrolls at the same time. Is this desired behavior?
The only way I can get to a certain freq at a certain zoom level is to Shift + Left Drag a marquee. Or am  I missing the idea? 
    On Wednesday, February 17, 2021, 01:29:31 AM PST, Chris Cannam <notifications@github.com> wrote:  
 
 


I can confirm this, thank you for the report.

However, it might not be all that bad - I've checked the code, and it looks as if the docs (i.e. the key and mouse reference) simply list Shift+Wheel and Alt+Wheel the wrong way around. Try Shift+Wheel to scroll vertically and Alt+Wheel to zoom.

I'm not sure whether it was always this way, or whether at some point the code has changed (but why?) to invalidate the docs. I suppose I could dig in and find out.

If you can confirm this works for you, then I'll go ahead and simply correct the docs for the next release.

(Most of the key combinations listed in Key and Mouse Reference are taken automatically from the actual menu shortcuts, and so are "guaranteed" to be current and correct - but the general pane navigation ones sadly are not)

**Comment by cannam** [comment]
Ah, I think I see. I think you're describing the behaviour you have when viewing a layer with a log frequency scale, like the melodic-range spectrogram.

The way vertical scrolling works in all layers is to move the centre point of the view but keep the visible range fixed within the unit of the data in the layer. In log-frequency spectrograms, that unit is Hz, and keeping the visible range in Hz constant while scrolling, when the vertical scale is set to logarithmic, means that the apparent range covered (in log-Hz or musical pitch) changes.

This is not an intentional behaviour as such, it's just a side-effect of the fact that vertical scrolling doesn't change the visible range, and the visible range in these layers happens to be measured using the underlying frequency scale rather than the logarithmic pitch one.

Does this sound like the situation you're observing?

**Comment by St3v3C** [comment]
Yes that is what I meant. Sorry I should have specified looking at a spectrogram. How I thought it should work is similar to a spectrum analyzer (display of logarithmic amplitude versus frequency) where the input gain sets the amplitude per division (analogous to zoom) and the attenuation sets the range displayed (analogous to scroll). So if one sets for example gain of 10dB per vertical division and then the displayed range to be 0 to -100dB, changing the range to -20 to -120 dB would not change the gain, still 10 dB per div in the vertical. Maybe it just doesn't work like that and I have a bad analogue!
Another way of saying that is that if in a Sonic Vizualizer spectrogram I set 440 Hz to 880 Hz to be 2 cm high on the display then scroll to move 440 Hz to the bottom of the display so I can see harmonics I would expect the distance on the display from 440 to 880 to still be 2 cm. Not the case. For me that makes it difficult to stay oriented on the display. So what I have to do instead is drag a marque around the area of interest then zoom way out and drag another marque each time I want to move in the vertical.
So a short word about my task... I'm picking out individual guitar notes in a clip that is recorded with bass and drums. I employed filtering on the clip to attenuate the drums in particular but the signal is very noisy so zooming and scrolling and staying oriented is a constant struggle. Sonic Vis is a great tool for this task though and I'm grateful to have it!
The second issue is that Alt + Wheel is purported to Scroll Vertically. For me it Scrolls _Horizontally_ that is Alt + Wheel does exactly the same thing as Ctrl + Wheel. They both only affect the horizontal on a spectrogram. 
    On Thursday, February 18, 2021, 02:26:02 AM PST, Chris Cannam <notifications@github.com> wrote:  
 
 


Ah, I think I see. I think you're describing the behaviour you have when viewing a layer with a log frequency scale, like the melodic-range spectrogram.

The way vertical scrolling works in all layers is to move the centre point of the view but keep the visible range fixed within the unit of the data in the layer. In log-frequency spectrograms, that unit is Hz, and keeping the visible range in Hz constant while scrolling, when the vertical scale is set to logarithmic, means that the apparent range covered (in log-Hz or musical pitch) changes.

This is not an intentional behaviour as such, it's just a side-effect of the fact that vertical scrolling doesn't change the visible range, and the visible range in these layers happens to be measured using the underlying frequency scale rather than the logarithmic pitch one.

Does this sound like the situation you're observing?

**Comment by cannam** [comment]
It's true that the behaviour you want simply isn't the way things work in SV - because the underlying model is fundamentally linear-Hz and all SV does is scroll it and then view the result on a log-Hz scale. I would certainly agree that, for use cases like this, the way SV does it is unhelpful. There is a root cause for this behaviour that makes some sense (it has to do with the way the underlying models are represented) but in this case it's seldom good from a user perspective. At this point, though, I'm afraid a wholesale change to that behaviour is unlikely.

I will certainly change the docs so that they have the shortcuts listed the correct way around (shift+wheel to scroll vertically [changing the zoom level as a side-effect, if you happen to have the scale set to logarithmic] and alt+wheel to scroll horizontally).

**Comment by cannam** [comment]
Oh, I've just realised that there is in fact a shortcut missing entirely (even in the context of SV's own way of thinking about things). The docs say

Wheel = zoom horizontally
Ctrl+wheel = scroll horizontally
Shift+wheel = zoom vertically
Alt+wheel = scroll vertically

The _comments_ in the code suggest that the intended behaviour is really

Wheel = zoom horizontally
Ctrl+wheel = scroll horizontally
Shift+wheel = scroll vertically (in an unsatisfactory way, when log-scaled)
Alt+wheel = zoom vertically

But the _actual_ behaviour is

Wheel = zoom horizontally
Ctrl+wheel = scroll horizontally
Shift+wheel = scroll vertically (in an unsatisfactory way, when log-scaled)
Alt+wheel = scroll horizontally

So something is getting lost - no shortcut zooms vertically, without scrolling.

It's quite possible that this got messed up with the introduction of support for trackpad gestures allowing horizontal two-finger scrolling - these come encoded as a horizontal wheel event, complicating the logic.

As an aside, from this distance it seems incredible that there is actually a shortcut attached to _double-clicking the middle mouse button_ - something that is practically impossible to do on any wheel mouse or touch surface! I'm not going to do anything with that now though.

**Comment by cannam** [comment]
It looks as if Qt itself is taking over the Alt modifier - when Alt is pressed, the wheel event is reported to us as horizontal rather than vertical. According to https://bugreports.qt.io/browse/QTBUG-30948 this behaviour was introduced in Qt 4.6 in 2009! (Although the [commit linked in that report](https://github.com/qt/qt/commit/f04af57ba71e8a4cd3c19f6e1a283290cc5280d4) does not actually appear to be the one that introduces it, I guess regardless this means the behaviour has been there for a long time and I have never noticed)

I think we probably ought to recognise the Alt modifier on horizontal scrolls and force them back to vertical, as it's the only way to get us both vertical actions without having to change what Ctrl means. I don't think we'll ever have a use for Alt+horizontal drag, so it's probably safe to make this change. It's presumably inconsistent with other applications that do expect alt-wheel to scroll horizontally, but SV's interaction has always been a bit weird anyway - "normal" behaviour would involve the wheel scrolling vertically and Ctrl+wheel zooming.

**Comment by cannam** [comment]
I've made that change, so the docs and the actual behaviour now match.

I'm aware that this doesn't fix the problem as you reported it, because that behaviour is by design, even if it is a misdesign! I'm going to change the bug title here and leave the report open, but to be honest I don't see much likelihood that this behaviour will change as it would be quite a deep thing to change at this point.

**Comment by St3v3C** [comment]
Thanks for looking into it and for the report!
    On Thursday, February 25, 2021, 02:59:42 AM PST, Chris Cannam <notifications@github.com> wrote:  
 
 


I've made that change, so the docs and the actual behaviour now match.

I'm aware that this doesn't fix the problem as you reported it, because that behaviour is by design, even if it is a misdesign! I'm going to change the bug title here and leave the report open, but to be honest I don't see much likelihood that this behaviour will change as it would be quite a deep thing to change at this point.
